File: solutions/go/simple-linked-list/1/simple_linked_list.go
package linkedlist

import "errors"

type Node struct {
	val  int
	next *Node
}

type List struct {
	head *Node
	tail *Node
}

func New(elements []int) *List {
	newList := List{}

	if len(elements) == 0 {
		return &newList
	}

	newList.head = &Node{
		val: elements[0],
	}
	curNode := newList.head

	for i := 1; i < len(elements); i++ {
		curNode.next = &Node{
			val: elements[i],
		}
		curNode = curNode.next
	}
	newList.tail = curNode

	return &newList
}

func (l *List) Size() int {
	size := 0

	for node := l.head; node != nil; node = node.next {
		size++
	}

	return size
}

func (l *List) Push(element int) {
	newNode := &Node{
		val: element,
	}
	if l.head == nil {
		l.head = newNode
		l.tail = newNode
	} else {
		l.tail.next = newNode
		l.tail = l.tail.next
	}
}

func (l *List) Pop() (int, error) {
	if l.head == nil {
		return 0, errors.New("cannot pop from empty list")
	} else if l.tail == l.head {
		temp := l.head.val
		l.head = nil
		l.tail = nil
		return temp, nil
	}

	node := l.head
	for ; node.next != l.tail; node = node.next {}

	temp := node.next.val
	l.tail = node
	l.tail.next = nil
	return temp, nil
}

func (l *List) Array() []int {
	newArray := make([]int, 0)

	for curNode := l.head; curNode != nil; curNode = curNode.next {
		newArray = append(newArray, curNode.val)
	}

	return newArray
}

func (l *List) Reverse() *List {
	var prev, next *Node
	current := l.head
	l.tail = l.head

	for current != nil {
		next = current.next
		current.next = prev
		prev = current
		current = next
	}

	l.head = prev
	return l
}

File: solutions/python/simple-linked-list/1/simple_linked_list.py
class Node:
    def __init__(self, value):
        self._value = value
        self._next = None

    def value(self):
        return self._value

    def next(self):
        return self._next


class LinkedList:
    def __init__(self, values=[]):
        self._head = Node(values[0]) if values else None
        for i in range(1, len(values)):
            self.push(values[i])

    def __len__(self):
        node = self._head
        length = 0

        while node is not None:
            node = node.next()
            length += 1

        return length

    def head(self):
        if len(self) == 0:
            raise EmptyListException("The list is empty.")
        return self._head

    def push(self, value):
        new_node = Node(value)
        new_node._next = self._head
        self._head = new_node

    def pop(self):
        to_return = self.head()
        self._head = self._head.next()
        return to_return.value()

    def __iter__(self):
        self._iter_next = self._head
        return self

    def __next__(self):
        if self._iter_next is None:
            raise StopIteration
        to_return = self._iter_next.value()
        self._iter_next = self._iter_next.next()
        return to_return

    def reversed(self):
        return reversed(list(self))


class EmptyListException(Exception):
    def __init__(self, message):
        self.message = message
